flang/lib/Lower/OpenMP/DataSharingProcessor.cpp

```diff
@@ -289,12 +290,34 @@ void DataSharingProcessor::collectSymbolsInNestedRegions(
        eval.getNestedEvaluations()) {
     if (nestedEval.hasNestedEvaluations()) {
       if (nestedEval.isConstruct())
-        // Recursively look for OpenMP constructs within `nestedEval`'s region
         collectSymbolsInNestedRegions(nestedEval, flag, symbolsInNestedRegions);
-      else
-        converter.collectSymbolSet(nestedEval, symbolsInNestedRegions, flag,
-                                   /*collectSymbols=*/true,
-                                   /*collectHostAssociatedSymbols=*/false);
+      else {
+        bool isOrderedConstruct = [&]() {
```

Comment:
@kiranchandramohan @luporl just to clarify this change:
Since `omp ordered` does not have its own data-sharing environment, I think we should consider symbols referenced by an `omp ordered` construct to be separate from the symbols referenced by the parent region.

The changes here make sure that for the following input, `i` to be collected as of the symbols to be privatized for the `omp do ordered` construct:
```fortran
  !$omp do ordered
  do i = 2, 10
    !$omp ordered
    a(i) = a(i-1) + 1
    !$omp end ordered
  end do
  !$omp end do
```

Reply:
It is not clear to me why the ordered construct is a special case. Will the same issue be resolved by a fix similar to https://github.com/llvm/llvm-project/pull/90671 ?

Reply:
I can describe the problem in more details but cannot claim my solution is the proper one.

For the above example program, this is the output of `flang-new -fc1 -fopenmp -fdebug-unparse-with-symbols`:
```fortram
!DEF: /foo (Subroutine) Subprogram
subroutine foo
 !DEF: /foo/a ObjectEntity INTEGER(4)
 !DEF: /foo/i ObjectEntity INTEGER(4)
 integer a(10), i
!$omp do  ordered
 !DEF: /foo/OtherConstruct1/i (OmpPrivate, OmpPreDetermined) HostAssoc INTEGER(4)
 do i=2,10
!$omp ordered 
  !REF: /foo/a
  !REF: /foo/OtherConstruct1/i
  a(i) = a(i-1)+1
!$omp end ordered 
 end do
!$omp end do 
end subroutine
```

So, `i` is marked as being `REF`ed from within the `omp ordered` construct (not defined).

However, when you call `converter.collectSymbolSet(nestedEval, symbolsInNestedRegions, ...)` when `nestedEval` is the `omp ordered` construct, `i` will be one of the symbols collected in that nested region.

This in turn, causes `i` **not** to be privatized when privatizing symbols that should be privatized for the `omp do ordered` construct (`i` is skipped because it is referenced by a nested region).

Since, `omp ordered` does not have a data environment of its own (as far as I understand), I thought it would reasonable to not exclude the symbols collected for a nested eval when that nested eval is an `ordered` directive.

That said, I am not sure this is the proper location for that kind of fix. But a similar fix to what was introduced in https://github.com/llvm/llvm-project/pull/90671 does not seem to work. Tried adding that in `bool OmpAttributeVisitor::Pre(const parser::OpenMPBlockConstruct &x);` but I get the same problem described above.

Reply:
I think we can fix this separately. Could you check whether the same issue happens with the Critical construct?

Reply:
https://github.com/llvm/llvm-project/pull/90671 fixed symbol handling in semantics, which was enough to fix https://github.com/llvm/llvm-project/issues/78936, but not https://github.com/llvm/llvm-project/issues/75767.

Maybe the issue here is similar to https://github.com/llvm/llvm-project/issues/75767.
In any case, I agree that this can be fixed separately.

Reply:
Thanks both for the review.

> Could you check whether the same issue happens with the Critical construct?

That indeed causes the same issue. I will add a similar "fix" and a test before merging.

Reply:
> I get semantic error as: `error: DEFAULT clause is not allowed on the DO directive`

I think this error happens earlier in semantics because `OMPC_Default` is not listed in the allowed clauses of `OMP_Do` in "llvm/include/llvm/Frontend/OpenMP/OMP.td". Not sure whether it should be allowed or not (making that error expected), since I'm not familiar with that part of the spec.

Reply:
Thanks! Indeed it is from semantics. I was mainly checking if default privatization logic is being invoked uncharacteristically when it should not (i.e. when `default(...)` is not specified).